walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
+
+    return free_time;
+  }
+
+  // This method checks if an event overlaps with any of our free time slots.
+  // It returns the conflicted slot's
+  private int containsEvent(ArrayList<TimeRange> free_time, TimeRange event_timerange) {
+    for (TimeRange slot : free_time) {
+      if (slot.overlaps(event_timerange)) return free_time.indexOf(slot);
+    }
+    return -1;
+  }
+
+  /**
+   * This method free time , conflicted free time slot and current event which we are looping
+   * through and checks all conditions in which they could overlap. Then returns the free time minus
+   * overlapping part.
+   */
+  private ArrayList<TimeRange> getNewFreeTime(
+      TimeRange free_slot,
+      TimeRange event_slot,
+      ArrayList<TimeRange> free_time,
+      int conflicted_slot_index) {
+
+    int free_slot_start = free_slot.start();
+    int free_slot_end = free_slot.end();
+
+    int event_slot_start = event_slot.start();
+    int event_slot_end = event_slot.end();
+
+    // Case: Free time: |---|
+    // Event slot:    |-------|
+    if (event_slot_start < free_slot_start && event_slot_end > free_slot_end) {
+      free_time.remove(free_slot);
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:     |-----|
+    // or
+    // |------|
+    // |--|
+    else if (event_slot_start <= free_slot_start && event_slot_end < free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(TimeRange.fromStartEnd(event_slot_end, free_slot_end, false)));
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:            |-----|
+    // or
+    // |------|
+    //     |--|
+    else if (event_slot_start > free_slot_start && event_slot_end >= free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(TimeRange.fromStartEnd(free_slot_start, event_slot_start, false)));
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:        |-----|
+    else if (event_slot_start > free_slot_start && event_slot_end < free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(
+              TimeRange.fromStartEnd(free_slot_start, event_slot_start, false),
+              TimeRange.fromStartEnd(event_slot_end, free_slot_end, false)));
+    }
+
+    // else return the original free time
+    return free_time;
+  }
+
+  /**
+   * This method checks the duration of each free time slot compares it to the requested meeting
+   * duration. If the free time slot can not accommodate the meeting duration then it removes that
+   * free time slot from the options (ie free time)
+   */
+  private ArrayList<TimeRange> checkDuration(ArrayList<TimeRange> free_time, int meeting_duration) {
+    for (int i = 0; i < free_time.size(); i++) {
+      if (free_time.get(i).duration() < meeting_duration) free_time.remove(i);
+    }
```

Comment:
Not a Java expert here, but this looks dangerous: You are removing items from a list you are iterating over.

So my understanding is, if your free_time has 3 elements and you remove the first, then you look at element 3 (re-indexed as 2) next. Since element 2 got re-indexed as 1 and you increment i every iteration it will simply be skipped.

Reply:
Oops sorry!
I completely overlooked that. 
I will fix it right now!
Thanks for letting me know.

Reply:
So, I'm trying to iterate it from the end. I think it should work now.